staging/kos/pkg/controllers/connectionagent/connectionagent.go

```diff
@@ -0,0 +1,1019 @@
+		// separate worker threads).
+		oldLocalAttRef := ca.fromAttToEnquableRef(oldLocalAtt)
+		ca.queue.Add(oldLocalAttRef)
+	}
+}
+
+func (ca *ConnectionAgent) onLocalAttRemoved(obj interface{}) {
+	peeledObj := peel(obj)
+	localAtt := peeledObj.(*netv1a1.NetworkAttachment)
+	glog.V(5).Infof("Notified of NetworkAttachment %#+v removal from local NetworkAttachments cache\n", localAtt)
+	localAttRef := ca.fromAttToEnquableRef(localAtt)
+	ca.queue.Add(localAttRef)
+}
+
+func (ca *ConnectionAgent) processLocalAtt(attRef attQueueRef) error {
+	localAtt, err := ca.localAttsLister.NetworkAttachments(attRef.nsn.Namespace).Get(attRef.nsn.Name)
+	if err != nil && !k8serrors.IsNotFound(err) {
+		// This should never happen.  No point in retrying.
+		glog.Errorf("NetworkAttachments Lister failed to lookup %s/%s: %s\n",
+			attRef.nsn.Namespace, attRef.nsn.Name, err.Error())
+		return nil
+	}
+	if err != nil && k8serrors.IsNotFound(err) {
+		// If we're here the attachment has been deleted
+		err = ca.releaseLocalAttResources(attRef)
+		return err
+	}
+	if err == nil && attRef.vni != localAtt.Spec.VNI {
+		// If we are here there's been an update on the vni field, and we must clear
+		// the resources associated with localAtt when it had the old vni field value.
+		err = ca.releaseLocalAttResources(attRef)
+		return err
+	}
+	err = ca.processExistingLocalAtt(localAtt)
+	return err
+}
+
+func (ca *ConnectionAgent) processRemoteAtt(attRef attQueueRef) error {
+	ifcKey := fromAttRefToVNIAndNsn(attRef)
+	stateOfAttVN, stateFound := ca.getVNStateForVNI(attRef.vni)
+	if !stateFound {
+		// If we're here the last local NetworkAttachment in the Virtual Network of the remote attachment
+		// referenced by attRef has been deleted => the attachment by attRef is no longer relevant for
+		// the connection agent.
+		deletedIfc, ifcFound, err := ca.deleteRemoteIfc(ifcKey)
+		if ifcFound {
+			glog.V(3).Infof("Deleted remote Network Interface %#+v\n", deletedIfc)
+		}
+		return err
+	}
+	remoteAtt, err := stateOfAttVN.remoteAttsLister.Get(attRef.nsn.Name)
+	if err != nil && !k8serrors.IsNotFound(err) {
+		// This should never happen. No point in retrying.
+		return nil
+	}
+	if err != nil && k8serrors.IsNotFound(err) {
+		// If we're here the remote attachment has been deleted.
+		deletedIfc, ifcFound, err := ca.deleteRemoteIfc(ifcKey)
+		if ifcFound {
+			glog.V(3).Infof("Deleted remote Network Interface %#+v\n", deletedIfc)
+		}
+		return err
+	}
+	return ca.createOrUpdateRemoteAttIfc(remoteAtt)
+}
+
+func (ca *ConnectionAgent) releaseLocalAttResources(attRef attQueueRef) error {
+	ca.updateVNStateForDeletedLocalAtt(attRef)
+	deletedIfc, ifcFound, err := ca.deleteLocalIfc(fromAttRefToVNIAndNsn(attRef))
+	if err == nil && ifcFound {
+		glog.V(3).Infof("Deleted local Network Interface %#+v\n", deletedIfc)
+	}
+	return err
+}
+
+// TODO break this method body into smaller methods, possibly shared with createOrUpdateRemoteAttIfc. Refactor it, make it more readable.
```

Comment:
See how most of the logic in `processEndpoint` at https://github.ibm.com/k8s4g/hermes-iobricks-network-workspace/blob/8f7169a53946b9f3d08a4556c8931fa50d980da4/src/github.ibm.com/genctl/network-controllers/pkg/connection-agent/agent.go#L2982 is common between local and remote.  Perhaps even more could be.